k-distribution/include/kframework/builtin/domains.md

```diff
@@ -1881,11 +1881,15 @@ of the above operations in K.
         requires findString(Source, ToReplace, 0) <Int 0
 
 
+  // Note that the replace function is undefined when Count < 0. This allows different backends to
+  // implement their own behavior without contradicting these semantics. For instance, a symbolic
+  // backend can return #Bottom for that case, while a concrete backend can throw an exception.
   rule replace(Source:String, ToReplace:String, Replacement:String, Count:Int) =>
        substrString(Source, 0, findString(Source, ToReplace, 0)) +String Replacement +String
        replace(substrString(Source, findString(Source, ToReplace, 0) +Int lengthString(ToReplace), lengthString(Source)), ToReplace, Replacement, Count -Int 1)
-        requires Count >Int 0
-  rule replace(Source:String, _, _, 0) => Source
+        requires Count >Int 0 andBool findString(Source, ToReplace, 0) >=Int 0
+  rule replace(Source:String, _, _, Count) => Source
+        requires Count >=Int 0 [owise]
```

Comment:
```suggestion
  rule replace(Source:String, _, _, 0) => Source [owise]
```

For the owise rule, the variant above will have proper behavior I think.

The version with the constraint `requires Count >= 0 [owise]` appears to make `replace` not uniquely defined because the term:

`replace(Source, ToReplace, Replacement, 1)`

Can evaluate into one of two possibilities:

1.   if `findString(Source, ToReplace, 0) >=Int 0`, then it can evaluate into:

     ```
     substrString(Source, 0, findString(Source, ToReplace, 0)) +String Replacement +String
       replace(substrString(Source, findString(Source, ToReplace, 0) +Int lengthString(ToReplace), lengthString(Source)), ToReplace, Replacement, Count -Int 1)
     ```
     
2.   also since `Count >=Int 0`, it can evaluate into:

     ```
     Source
     ```
     
If we're all on the same page, then I'm happy to approve after the patch is applied.

Reply:
Because it's an owise rule though, it should always evaluate to case 1, unless the findString condition doesn't hold, then it will evaluate to case 2.

What should `replace` do if `Count >Int 0` and `findString(...) ==Int -1`? I believe your proposed solution makes it undefined.

Reply:
You're absolutely right! Thanks for fact-checking me.